compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```
@@ -649,29 +649,334 @@ abstract class GenJSCode extends plugins.PluginComponent
+        dispatchResolution match {
+          case js.Block(stats) =>
+            val js.ApplyStatic(_, method, _) = stats.last
+            val refs = constructorTree.getParamRefs(method.name)
+            val rhss = stats.collect {
+              case js.VarDef(_, _, _, rhs) => rhs
+            }
+            zipMap(refs, rhss) { (ref, rhs) =>
+              js.VarDef(ref.ident, ref.tpe, mutable = false, rhs)
+            }
+
+          case js.ApplyStatic(_, mtd, js.This() :: Nil)
+              if ir.Definitions.isConstructorName(mtd.name) =>
+            Nil
+
+          case tree =>
+            List(tree)
+        }
+      } else {
+        val overloadRef = js.VarRef(overloadIdent)(jstpe.IntType)
+
+        def transformDispatch(tree: js.Tree): js.Tree = tree match {
+          case js.Block(stats) =>
+            val js.ApplyStatic(_, method, _) = stats.last
+            val num = constructorTree.getOverrideNum(method.name)
+            val refs = overloadRef :: constructorTree.getParamRefs(method.name)
+            val rhss = js.IntLiteral(num) :: stats.collect {
+              case js.VarDef(_, _, _, rhs) => rhs
+            }
+            val newStats = zipMap(refs, rhss)(js.Assign(_, _))
+            js.Block(newStats)
+
+          case js.ApplyStatic(_, method, js.This() :: Nil)
+              if ir.Definitions.isConstructorName(method.name) =>
+            js.Assign(overloadRef,
+                js.IntLiteral(constructorTree.getOverrideNum(method.name)))
+
+          case js.Match(selector, cases, default) =>
+            val newCases = cases.map {
+              case (literals, body) => (literals, transformDispatch(body))
+            }
+            val newDefault = transformDispatch(default)
+            js.Match(selector, newCases, newDefault)(tree.tpe)
+
+          case js.If(cond, thenp, elsep) =>
+            js.If(cond, transformDispatch(thenp),
+              transformDispatch(elsep))(tree.tpe)
+
+          case _ =>
+            tree
```

Can this happen? It seems this is asking for silently ignoring cases where we don't have the shape we expected.